server/utils/controllerUtils.ts

```diff
@@ -10,4 +24,35 @@ export default class ControllerUtils {
 
     res.render(...view.renderArgs)
   }
+
+  static async fetchDraft<T>(
+    req: Request,
+    res: Response,
+    draftsService: DraftsService,
+    {
+      idParamName,
+      notFoundUserMessage,
+      typeName,
+    }: { idParamName: string; notFoundUserMessage: string; typeName: string }
+  ): Promise<DraftFetchResult<T>> {
+    const id = req.params[idParamName]
+    const draft = await draftsService.fetchDraft<T>(id, {
+      userId: res.locals.user.userId,
+    })
+
+    if (draft === null) {
+      throw createError(500, `Draft ${typeName} with ID ${id} not found by drafts service`, {
+        userMessage: notFoundUserMessage,
+      })
+    }
+
+    if (draft.softDeleted) {
```

Comment:
I think that rendering a View should not be the responsibility of this method and instead should be the caller's responsibility. 
This is because:
1. The caller shouldn't expect the response to be terminated based on fetching a draft.
2. The caller will have to explicitly put scaffolding after calling to ensure that softDeleted drafts don't break the reset of the request. If you're doing that may as well do the rendering as well.

What do you think about throwing a custom error which is caught by the errorHandler to render the specific page?

Reply:
I think they're valid concerns, but I wonder if it could perhaps be solved by improved naming of the method, like `fetchDraftOrRenderMessage`? I think it's possibly fair game for a method in `ControllerUtils `to render something, if we think it'd be reasonable for a helper instance method in a controller to do the same.

It's true that the caller could instead render the message, given that it has to check whether a render has happened anyway. My worry is that every controller method that wants to fetch a draft would then also have to make sure they check whether the draft is soft deleted before allowing it to be used. The way that the method is written at the moment, either the `fetchDraft` method returns a usable draft, or it doesn't return a draft at all. But I'm happy to move the rendering to the caller if you think the current approach is particularly egregious?

I'm also not a big fan of using exceptions to modify behaviour and control flow when there are more explicit ways to do it. I think that things like middlewares can be useful but they can also obscure how a controller method behaves by making it the result of multiple behaviours that are distant from each other in the codebase. I think that if we went down the route of using the `errorHandler` to handle a growing number of error-_ish_ cases then that class would become larger and larger as we started to handle more special cases. Unless we built a mechanism where we could attach an arbitrary view to the error for the error handler to render. But I think I'm a bit ambivalent about going down that route.

Reply:
> My worry is that every controller method that wants to fetch a draft would then also have to make sure they check whether the draft is soft deleted before allowing it to be used.

I personally think that is preferable than having the method rendering a page without the caller knowing about it. It is not obvious when looking at the code. Perhaps renaming the method might be best solution... 

I see your point about the caller having to know what to do with a softDeleted draft; however it can still be considered the callers responsibility to deal with it how they feel best - perhaps they might want to render what has been soft deleted, or perhaps the time in which it was soft deleted, etc.

One suggestion is to change the logic to this:
```
   const draft = await this.fetchDraftBookingOrThrowSpecificError(req, res)
    if (draft.softDeleted) {
         controllerUtils.renderDraftSoftDeletedView(req, res)
    }
```

Functionally it's no different than what you have currently, but this makes it more obvious on the controller level how pages are rendered based on logic.

Reply:
I think your suggestion is a good one too. And if it _does_ turn out that we'd like to have custom handling for soft deleted drafts (or custom logic for which drafts to disallow), then can switch to that pattern in the future. In the interests of getting this released and not being sure whether we need that kind of customisability yet, I think that I'd like to stick with the working solution that I've already done. I've renamed the methods to hopefully make it clearer what they do. Please could you take another look?